### PR TITLE
fix kraken bug, allow for live order edits

### PR DESCRIPTION
**Pull request by iamzoltan**
This seems to work. I tested it a bunch, but if @goodboy, @guilledk, or @wattygetlood could also test it out. That would be great.

I can't believe I missed this the first time hah.

### CODE REVIEW

**Comment by iamzoltan** [comment]
this address #299, some CI tests fail, not sure why

**Comment by goodboy** [comment]
CI is annoying new security stuff to do with git.

File: piker/brokers/kraken.py [hunk 1]
```diff
@@ -313,29 +305,41 @@ async def submit_limit(
...
+                "validate": False
+            }
+            return await self.endpoint('AddOrder', data)
+        else:
+            # Edit order data for kraken api
+            data = {
+                "txid": reqid,
+                "pair": symbol,
+                "price": str(price),
+                # set to True test EditOrder call without a real submission
```
Comment (goodboy) [inline piker/brokers/kraken.py:332]: Hm, can we expose this up through the method as kwarg too maybe?

Both endpoints support it so i think just a wholesale flag would be handy here if we try to implement a pseudo-paper engine no?

File: piker/brokers/kraken.py [hunk 2]
```diff
@@ -313,29 +305,41 @@ async def submit_limit(
...
-            "volume": str(size),
-            "pair": symbol,
-            "price": str(price),
-            # set to True test AddOrder call without a real submission
-            "validate": False
-        }
-        return await self.endpoint('AddOrder', data)
+        if reqid is None:
+            # Build order data for kraken api
+            data = {
```
Comment (goodboy) [inline piker/brokers/kraken.py:316]: the `data: dict` duplication isn't necessary right?

in the edit case can't you just update with the `txid` ?

**Review by goodboy — COMMENTED** [review]
Left a couple quiffs but nothing deal breaking.

I'm going to test this now.

**Comment by goodboy** [comment]
@iamzoltan oh only last thing would be disabling the config safeguard i put in yah?

File: piker/brokers/kraken.py [hunk 3]
```diff
@@ -161,13 +160,6 @@ def get_config() -> dict[str, Any]:
     conf, path = config.load()
     section = conf.get('kraken')
 
-    if section:
```
Comment (goodboy) [inline piker/brokers/kraken.py:164]: wait nm i missed this..

**Review by guilledk — APPROVED** [review]

File: piker/brokers/kraken.py (hunk 3 above)
Comment (iamzoltan) [reply piker/brokers/kraken.py:164]: what do you meam?

File: piker/brokers/kraken.py (hunk 2 above)
Comment (iamzoltan) [reply piker/brokers/kraken.py:316]: you mean, create the `dict` with the common keys from both end points, and then depending on which one we are using, add the extras?

File: piker/brokers/kraken.py (hunk 1 above)
Comment (iamzoltan) [reply piker/brokers/kraken.py:332]: not sure what you mean by this, can you elaborate?

File: piker/brokers/kraken.py [hunk 4]
```diff
@@ -596,16 +597,21 @@ async def handle_order_requests(
...
                 )
             else:
-                # TODO: handle multiple cancels
+                # TODO: handle multiple orders (cancels?)
                 #       txid is an array of strings
-                reqid = resp['result']['txid'][0]
+                if order.reqid is None:
+                    reqid = resp['result']['txid'][0]
+                else:
+                    # update the internal pairing of oid to krakens
```
[the remark this answers is not in the document]
Comment (iamzoltan) [reply piker/brokers/kraken.py:612]: basically, if its the first time the ems has seen the order, then the `reqid` is none, until we get a response from kraken. The the `txid` from kraken populates the `reqid` in the `ack` message.

then if the order is changed, and a `reqid` exists, the `EditOrder` endpoint gets pinged, and we need to change the `reqid` in the `ack` message to be associated with the new `txid` that is returned by kraken

File: piker/brokers/kraken.py (hunk 1 above)
Comment (goodboy) [reply piker/brokers/kraken.py:332]: the commented parameter that seems to make this a paper-like request?

File: piker/brokers/kraken.py (hunk 2 above)
Comment (goodboy) [reply piker/brokers/kraken.py:316]: yes

File: piker/brokers/kraken.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (goodboy) [reply piker/brokers/kraken.py:612]: right so i think the answer to my question is just *yes*?

File: piker/brokers/kraken.py (hunk 3 above)
Comment (goodboy) [reply piker/brokers/kraken.py:164]: don't worry about it

File: piker/brokers/kraken.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (iamzoltan) [reply piker/brokers/kraken.py:612]: yes?

File: piker/brokers/kraken.py (hunk 2 above)
Comment (iamzoltan) [reply piker/brokers/kraken.py:316]: aight

File: piker/brokers/kraken.py (hunk 1 above)
Comment (iamzoltan) [reply piker/brokers/kraken.py:332]: yes I hear you, as an arg in the `submit_limit` method, on it

**Review by iamzoltan — COMMENTED** [review]
thanks for the comments @goodboy

ill get these pushes up asap

File: piker/brokers/kraken.py [hunk 5]
```diff
@@ -308,34 +300,38 @@ async def get_trades(
...
             "pair": symbol,
             "price": str(price),
-            # set to True test AddOrder call without a real submission
-            "validate": False
+            "validate": validate
         }
-        return await self.endpoint('AddOrder', data)
+        if reqid is None:
+            # Build order data for kraken api
+            data["ordertype"] = "limit"
```
Comment (goodboy) [inline piker/brokers/kraken.py:322]: fwiw this is a good spot to use the new merge syntax:

`data |= {'ordertype': 'limit', 'type': action, 'volume': str(size)}`

if you wanted but nbd and it's def not a requirement for this pr.

Comment (goodboy) [reply piker/brokers/kraken.py:322]: oh and i also prefer the name `payload` for inputs to endpoints just for the next time.

**Review by goodboy — APPROVED** [review]
Good enough for me and I tested and everything lookin right.

Nice work @iamzoltan 🏄🏼

File: piker/brokers/kraken.py (hunk 5 above)
Comment (iamzoltan) [reply piker/brokers/kraken.py:322]: oh slick, I like that new syntax, I will update it. I can also switch data to payload if that is consistent with the rest of the base

**Review by iamzoltan — COMMENTED** [review]
gonna update that syntax, test it out, push and then merge. Thanks for the comments

File: piker/brokers/kraken.py (hunk 5 above)
Comment (iamzoltan) [reply piker/brokers/kraken.py]: acutally given ill leave the payload change for next time. <3

**Pull request merged**